File: hacktm/obey_the_rules/bpf.c
#include <sys/prctl.h>
#include <sys/socket.h>
#include <sys/syscall.h>
#include <sys/types.h>

#include <linux/audit.h>
#include <linux/filter.h>
#include <linux/seccomp.h>

#define Allow(syscall)                                                         \
  BPF_JUMP(BPF_JMP + BPF_JEQ + BPF_K, SYS_##syscall, 0, 1),                    \
      BPF_STMT(BPF_RET + BPF_K, SECCOMP_RET_ALLOW)

/*
struct sock_filter {	 Filter block
        __u16	code;    Actual filter code
        __u8	jt;	 Jump true
        __u8	jf;	 Jump false
        __u32	k;       Generic multiuse field
};
*/

struct sock_filter filter[] = {
    /* list of allowed syscalls */
    Allow(exit_group), /* exits a processs */
    Allow(brk),        /* for malloc(), inside libc */
    Allow(mmap),       /* also for malloc() */
    Allow(munmap),     /* for free(), inside libc */
    Allow(write),      /* called by printf */
    Allow(fstat),      /* called by printf */
};
struct sock_fprog filterprog = {.len = sizeof(filter) / sizeof(filter[0]),
                                .filter = filter};

File: hacktm/obey_the_rules/bpf.py
class BPFInsn:  # Filter block
    def __init__(self):
        self.code = None  # Actual filter code, word
        self.jt = None  # Jump true, byte
        self.jf = None  # Jump false, byte
        self.k = None  # Generic multiuse field, double-word

    def __repr__(self):
        return str((self.code, self.jt, self.jf, self.k))


def create():
    sock_filter = [BPFInsn() for _ in range(1)]
    sock_filter[0].code = 6
    sock_filter[0].jt = 0
    sock_filter[0].jf = 0
    sock_filter[0].k = 0x7FFF0000
    write_bpf(sock_filter, "bpf.bin")

    import struct

    open("bpf.bin", "wb").write(
        "".join(struct.pack("<HBBI", c.code, c.jt, c.jf, c.k) for c in sock_filter)
    )


def parse():
    rules = "2000000004000000"
    rules += "1500000a3e0000c0"
    rules += "2000000000000000"
    rules += "3500080000000040"
    rules += "1500060002000000"
    rules += "1500050038000000"
    rules += "1500000500000000"
    rules += "2000000010000000"
    rules += "1500000203000000"
    rules += "2000000018000000"
    rules += "1500000188286000"
    rules += "060000000000ff7b"
    rules += "0600000000000000"
    rules += "0000000000000000"
    rules += "0000000000000000"
    rules += "0000000000000000"
    rules += "0000000000000000"
    rules += "0000000000000000"

    sock_filter = []
    for i in range(0, len(rules), 16):
        sf = BPFInsn()
        sf.code = int(rules[i : i + 4], 16)
        sf.jt = int(rules[i + 4 : i + 6], 16)
        sf.jf = int(rules[i + 6 : i + 8], 16)
        sf.k = int(rules[i + 8 : i + 16], 16)
    open("server_bpf.bin", "wb").write(rules.decode("hex"))


parse()
